File: hw1/scala/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.json4s.jackson.JsonMethods.{compact, parse}

import java.io.{File, PrintWriter}
import scala.io.Source

object task1 {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf().setMaster("local[*]").setAppName("task1")
    val sc = new SparkContext(conf)
    val input_file = args(0)//"E:\\Data\\inf553\\hw1\\review.json"
    val output_file = args(1)//"task1_ans"
    val stopwords = args(2)//"E:\\Data\\inf553\\hw1\\stopwords"
    val y = args(3)//"2018"
    val m = args(4)//"10"
    val n = args(5)//"10"
    val rdd = sc.textFile(input_file).map(item => parse(item))
    //    val result_a = rdd.count()
    val rdd_b = rdd.filter(item => compact(item \ "date").contains(y))
    //    val result_b = rdd_b.count()
    val rdd_c = rdd.map(item => compact(item \ "user_id")).distinct()
//    val result_c = rdd_c.count()
    val rdd_d = rdd.map(item => (compact(item \ "user_id"), 1)).reduceByKey((a, b) => a + b)
      .sortBy(item => item._1).sortBy(item => item._2, ascending = false)
//      .map(item => List(item))
//    val result_d = rdd_d.take(m.toInt)
    val rdd_sw = sc.textFile(stopwords)
    var sw = Source.fromFile(stopwords).getLines().toList
    val punctuations = List("(", "[", ",", ".", "!", "?", ":", ";", "]", ")")
    for (p <- punctuations) {
      sw = sw :+ p
    }
    sw = sw :+ ""
    val rdd_e = rdd.flatMap(item => compact(item \ "text").split(" "))
      .map(item => item.trim().toLowerCase()).filter(item => !sw.contains(item))
      .map(item => (item, 1)).reduceByKey((a, b) => a + b)
      .sortBy(item => item._1).sortBy(item => item._2, ascending = false)
      .map(item => item._1)

    val pw = new PrintWriter(new File(output_file))
    pw.write("{\"A\": "+rdd.count()+", ")
    pw.write("\"B\": "+rdd_b.count()+", ")
    pw.write("\"C\": "+rdd_c.count()+", ")
    val r_d = rdd_d.take(m.toInt).toList
    pw.write("\"D\": [")
    r_d.foreach(item=>{
      pw.write("["+item._1+", "+item._2+"]")
      if (item._1 != r_d.last._1){
        pw.write(", ")
      }
    })
    pw.write("] \"E\": "+rdd_e.take(n.toInt).toList.mkString("[", ", ", "]")+"}")
    pw.close()

  }
}

File: hw1/python/task1.py
import sys
import json
from pyspark import SparkContext

def test(rdd):
    print(rdd.take(4))

def t1_a(rdd):
    return rdd.count()


def t1_b(rdd, y):
    rdd1 = rdd.filter(lambda x: y.strip() in x['date'])
    # return rdd1.take(4)
    return rdd1.count()

def t1_c(rdd):
    rdd1 = rdd.map(lambda x: x['user_id']).distinct()
    return rdd1.count()

def t1_d(rdd, m):
    rdd1 = rdd.map(lambda x: (x['user_id'], 1)).reduceByKey(lambda a, b: a+b).map(lambda x: list(x))
    rdd2 = rdd1.sortBy(lambda x: x[0]).sortBy(lambda x: x[1], ascending=False)
    return rdd2.take(m)

def t1_e(rdd, stop_words, n):
    punctuations = ["(", "[", ",", ".", "!", "?", ":", ";", "]", ")"]
    stop_words.append("")
    for p in punctuations:
        stop_words.append(p)
    rdd1 = rdd.flatMap(lambda x: x['text'].split(' ')).map(lambda x: x.lower().strip()).filter(lambda x: x not in stop_words)
    rdd2 = rdd1.map(lambda x: [x, 1]).reduceByKey(lambda a, b: a+b)
    rdd3 = rdd2.sortBy(lambda x: x[0]).sortBy(lambda x: x[1], ascending=False).map(lambda x:x[0])
    return rdd3.take(n)

if __name__ == '__main__':

    # input_file, output_file, stopwords, y, m, n = "review.json", "task1_ans", "stopwords", "2018", 10, 10
    input_file, output_file, stopwords, y, m, n = sys.argv[1:]
    m, n = int(m), int(n)
    sc = SparkContext("local[*]", 'task1')
    rdd_review = sc.textFile(input_file).map(json.loads)

    stop_words = sc.textFile(stopwords).collect()
    # print(t1_e(rdd_review, stop_words, n))
    result = dict()
    result['A'] = t1_a(rdd_review)
    result['B'] = t1_b(rdd_review, y)
    result['C'] = t1_c(rdd_review)
    result['D'] = t1_d(rdd_review, m)
    result['E'] = t1_e(rdd_review, stop_words, n)
    with open(output_file, 'w') as f:
        json.dump(result, f)
    # test(rdd_review)


